houdini/handlers/games/ninja/card.py

```diff
@@ -238,24 +236,31 @@ async def ninja_rank_up(p, ranks=1):
             await p.add_inbox(p.server.postcards[CardJitsuLogic.PostcardAwards[rank]])
         if rank in CardJitsuLogic.StampAwards:
             await p.add_stamp(p.server.stamps[CardJitsuLogic.StampAwards[rank]])
-    await p.update(ninja_rank=p.ninja_rank + ranks, ninja_progress=0).apply()
+    await p.update(ninja_rank=p.ninja_rank + ranks).apply()
     return True
 
+def get_exp_difference_to_next_rank(cur_rank: int) -> int:
+    return (cur_rank + 1) * 5
+
+def get_threshold_for_rank(rank: int) -> int:
+    # using arithmetic progression sum because the exp structure allows 
+    return (rank + 1) * rank // 2 * 5
 
 async def ninja_progress(p, won=False):
-    if p.ninja_rank == 0:
-        await p.update(ninja_progress=100).apply()
-    elif p.ninja_rank < 9:
-        speed = type(p.waddle).RankSpeed
-        if not won:
-            speed *= 0.5
-        points = math.floor((100 / p.ninja_rank) * speed)
-        await p.update(ninja_progress=p.ninja_progress+points).apply()
-    if p.ninja_progress >= 100:
+    # black belts don't need exp, otherwise it could overflow
+    if p.ninja_rank >= 9:
+        return
+    gained_exp = 5 if won else 1
+    next_rank_threshold = get_threshold_for_rank(p.ninja_rank + 1)
+
+    # this clamping is for compatibility with older versions
+    previous_progress = max(min(p.ninja_progress, next_rank_threshold), get_threshold_for_rank(p.ninja_rank))
```

Comment:
I think maybe, we should still approximate the amount of XP based on the previous progress?

Reply:
so for example, if in the old data they only have the white belt + 90% progress it wouldn't make them jump ranks immediately?

Reply:
OH nvm but still if they had 50% progress it would still jump them to the next rank after a game